File: go/rate_limiter.go
package ratelimit

// UserTier is "free" or "paid".
type UserTier string

const (
	Free UserTier = "free"
	Paid UserTier = "paid"
)

// UserTiers maps user ID to tier. Tests set this; the rate limiter reads it.
var UserTiers = make(map[string]UserTier)

// RateLimiter returns true if the request is allowed, false if it is rate limited.
//
// TODO: implement per SPEC.md.
//   - Look up the user's tier in UserTiers.
//   - Free: allow the first FREE_LIMIT requests ever, then always deny.
//   - Paid: allow at most PAID_LIMIT requests per WINDOW_SECONDS-second window.
//
// Keep per-user state in memory (e.g. package-level maps).
func RateLimiter(userID string) bool {
	// Stub: replace with your implementation.
	_ = userID
	return false
}

File: csharp/RateLimiter.cs
using System.Collections.Concurrent;

namespace ApiRateLimiter;

/// <summary>
/// API rate limiter.
/// <list type="bullet">
///   <item>Free users: <see cref="Constants.FreeLimit"/> requests total, ever (lifetime cap; never resets).</item>
///   <item>Paid users: <see cref="Constants.PaidLimit"/> requests per <see cref="Constants.WindowSeconds"/>-second window.</item>
/// </list>
/// Implement <see cref="AllowRequest"/> below. Everything else is scaffolding used by the tests.
/// (The method is not named RateLimiter because a member cannot share its enclosing type's name.)
/// </summary>
public static class RateLimiter
{
    /// <summary>Map from user ID to tier. Tests set this; the rate limiter reads it.</summary>
    public static readonly ConcurrentDictionary<string, UserTier> UserTiers = new();

    /// <summary>
    /// Returns true if the request is allowed, false if it is rate limited.
    ///
    /// TODO: implement per SPEC.md.
    ///   - Look up the user's tier in <see cref="UserTiers"/>.
    ///   - Free: allow the first FreeLimit requests ever, then always deny.
    ///   - Paid: allow at most PaidLimit requests per WindowSeconds-second window.
    /// Keep per-user state in memory (e.g. static dictionaries in this class).
    /// </summary>
    public static bool AllowRequest(string userId)
    {
        // Stub: replace with your implementation.
        return false;
    }
}
